### DISCUSSION TITLE
withAuthentication(String, CharSequence) usage in sentinel authentication ?

### DISCUSSION THREAD

**Discussion by Pramit2512** [redis/lettuce#2124] [Q&A] [answered]
I recently noticed in lettuce docs that below usage is deprecated

`sentinel(String host, int port, String masterId, CharSequence password)`
Deprecated. since 6.0,
 **use `sentinel(String, int, String)` and `withAuthentication(String, CharSequence`) instead.**

and when i check `
public RedisURI.Builder withAuthentication(String username, CharSequence password)
`

I am not able to understand the **String username** in parameter list for method **withAuthentication** . What is it ,and how can we get this/configure this .Is this any random string or sentinel username .if so how can we set/get  the sentinel username from config file.
I have never seen username usage before.Can anyone please explain?

**Comment by mp911de** [comment]
With Lettuce 6.0, we introduced support for Redis 6 ACL's. The correct way to setup sentinel with authentication is `RedisURI.Builder.sentinel(host, port).withPassword(…)` or `RedisURI.Builder.sentinel(host, port).withAuthentication(username, password)`.

**Comment by Pramit2512** [comment]
@mp911de , I have already tried using `RedisURI.Builder.sentinel(host, port).withPassword(…)` but i am still getting **Caused by: io.lettuce.core.RedisCommandExecutionException: NOAUTH Authentication required**.

My sentinel and data nodes both are password protected .Below is my connection code.It works when i have setup the master password only in master config file .As soon as i introduce sentinel password also ,it fails.


```
`  RedisURI.Builder
                        .sentinel(getSentinelHost_1(), getSentinelPort_1())
                        .withPassword(getSentinelPassword())
                        .withSentinel(getSentinelHost_2(), getSentinelPort_2())
                        .withSentinel(getSentinelHost_3(), getSentinelPort_3())
                        .withSentinelMasterId(getMasterName()).withPassword(getMasterPassword())
                        .build())`
```

Note::I am able to connect to redis if i am using deprecated version of sentinel authentication 

```
  RedisURI.Builder
                       .sentinel(getSentinelHost_1(), getSentinelPort_1(), getMasterName(), getSentinelPassword())
                        .withSentinel(getSentinelHost_2(), getSentinelPort_2())
                        .withSentinel(getSentinelHost_3(), getSentinelPort_3())
                        .withSentinelMasterId(getMasterName()).withPassword(getMasterPassword())
                        .build()
```

Can you please tell if i am missing something ,or is there a different way of doing it. 

Also for `RedisURI.Builder.sentinel(host, port).withAuthentication(username, password)`  I am not able to figure out what is username and password name as specified in its method definition.

**Comment by mp911de** [comment, marked as answer]
There's a difference between sentinel authentication and data node authentication. I think you need to post-process the `RedisURI` object after creation. `withPassword`/`withAuthentication` configures data node authentication. You can set authentication on the individual sentinels through:

```
for (RedisURI sentinel : redisURI.getSentinels()) {
    sentinel.setUsername()/sentinel.setPassword();
}
```

**Reply by Pramit2512** [reply]
Hello @mp911de (Mark),

With the latest lettuce 6.2.0 even the above approach is deprecated. Can you please suggest any other approach if you have idea.
can we use `withAuthentication(String, CharSequence)` for  sentinel password.

**Comment by Pramit2512** [comment]
Thansk a lot @mp911de  , it worked . Final code looks like this:

```
    RedisClient redisClient = RedisClient.create();

        LOG.info("created the redis client");

        RedisURI source = RedisURI.Builder
                .sentinel(getSentinelHost_1(), getSentinelPort_1())
                .withSentinel(getSentinelHost_2(), getSentinelPort_2())
                .withSentinel(getSentinelHost_3(), getSentinelPort_3())
                .withSentinelMasterId(getMasterName()).withPassword(getMasterPassword())
                .build();

        for (RedisURI sentinel : source.getSentinels()) {
            sentinel.setPassword(getSentinelPassword());
        }

        StatefulRedisMasterReplicaConnection<String, String> connection = MasterReplica.connect(redisClient, StringCodec.UTF8,
                source);

        connection.setReadFrom(ReadFrom.REPLICA_PREFERRED);

        LOG.info("created the redis connection");
```
